### ISSUE TITLE
Libssl/libcrypto support

### ISSUE THREAD

**Issue by skelsec** [pyodide/pyodide#2124]
## 🚀 Feature

Implement the python "ssl" module, or otherwise provide libssl/libcrypto support preferably via ctypes (or CFFI).

### Motivation

I'm currently working on a larger project that involves porting multiple network libraries to pyodide including SMB, LDAP, KERBEROS, RDP. Network stack is provided via a custom websocket proxy, but could be integrated with already existing WS-to-TCP proxy implementations.
All modules have a working prototype, but without SSL support these libraries are currently limited in capability: Kerberos can't support certificate based auth, LDAP can't support LDAPS, RDP only supports the RC4 based crypto etc.

### Pitch

There were multiple issues created on this matter, mostly mentioning that the current state of pyodide lacking performant crypto library support. Providing libssl/libcrypto would tackle this.  

What I see there are two possible solutions:
 - have a full `ssl` module implementation and one selected 3rd party library for generic crypto operations (would propose `pycryptodome`)
 - compile and package `openssl` as a shared library and have users interface with it via ctypes

### Alternatives

I see no alternatives to this. Most functionality provided `libcrypto` could be implemented in pure python but with severe performance penalty. `libssl` on the other hand is not something that can be easily bypassed.

### Additional context

I have tried to compile `openssl 1.1.1m` and adding it to `oscrypto`. So far I was only successful in compiling libcrypto in a way that could be loaded via `ctypes` and performed an AES-CTR operation without errors. But using the `libss.so` from the same run yields `malloc error` upon importing the module (in FireFox, Chrome refuses to load it).  
I have tried to compile `cryptography` but got stuck tackling the Rust requirement. Using a version which doesn't require Rust yields openssl compilation errors.
I have tried to compile `pyopenssl` but that requires `cryptography`
I have tried to compile `pycryptodome` whcih did not work, but I don't even remember what the error was.  
My current crypto "solution" is that I added a pure-python backend to `oscrypto` which allows me to perform most cryptographic operations (including pkcs1, parsing pem, rsa sign/decrypt, aes, rc2/4 etc) but obviously no SSL.

**Comment by hoodmane** [comment]
As far as I am aware, the problem with ssl is the sockets. It isn't allowed for programs running inside of a browser to directly touch a socket. Are the other features in ssl useful without sockets? We could probably compile a version of ssl that just raises an error if it tries to touch a real socket.

**Comment by skelsec** [comment]
Yes, basically with `ssl` you can wrap any streaming data, it doesn't have to be a socket. I wrote a wrapper around an async queue so I can deal with the "how" I'm getting the data there. Below is the actual class I'm using to have a transport-independent `SSL` implementation on `asyncio`.   
If you could add this functionality it would be greatly appreciated 

```
import asyncio

import platform
import traceback
try:
	import ssl
except:
	if platform.system() == 'Emscripten':
		#pyodide doesnt support SSL for now.
		pass


import logging
from asn1crypto.x509 import Certificate, PublicKeyInfo

class SSLClientTunnel:
	def __init__(self):
		self.in_queue = None
		self.out_queue = None
		self.ssl_ctx = None
		self.transport = None
		self.ssl_in_buff = None
		self.ssl_out_buff = None
		self.ssl_obj = None
		self.__ssl_in_q = None
		self.stop_evt = None
		self.server_certificate = None

		self.__enc_task = None
		self.__dec_task = None
		self.__process_task = None
	
	async def disconnect(self):
		try:
			if self.__enc_task is not None:
				self.__enc_task.cancel()
			if self.__dec_task is not None:
				self.__dec_task.cancel()
			if self.__process_task is not None:
				self.__process_task.cancel()

			if self.transport is not None:
				self.transport.out_queue.put_nowait(b'')
				await self.transport.disconnect()
		except Exception as e:
			pass
	
	async def get_server_pubkey(self):
		try:
			if self.server_certificate is None:
				return None
			cert = Certificate.load(self.server_certificate)
			return cert['tbs_certificate']['subject_public_key_info']['public_key'].dump()[5:] #why?
		except Exception as e:
			return e
	
	async def setup(self):
		try:
			self.in_queue = asyncio.Queue()
			self.out_queue = asyncio.Queue()
			self.stop_evt = asyncio.Event()
			self.__ssl_in_q = asyncio.Queue()
			self.ssl_in_buff = ssl.MemoryBIO()
			self.ssl_out_buff = ssl.MemoryBIO()
			self.ssl_obj = self.ssl_ctx.wrap_bio(self.ssl_in_buff, self.ssl_out_buff, server_side=False, server_hostname = None)
			_, err = await self.__do_handshake()
			if err is not None:
				raise err
			#print('Handshake ok!')
			self.server_certificate = self.ssl_obj.getpeercert(binary_form=True)
			self.__enc_task = asyncio.create_task(self.__encrypt_ssl_rec())
			self.__dec_task = asyncio.create_task(self.__decrypt_ssl_rec())
			self.__process_task = asyncio.create_task(self.__read_ssl_record())
			return True, None

		except Exception as e:
			return None, e
	
	async def __encrypt_ssl_rec(self):
		try:
			while True:
				try:
					raw_data = await self.out_queue.get()
					if raw_data == b'':
						await self.in_queue.put(None)
						#print('Connection terminated')
						return
					
					self.ssl_obj.write(raw_data)
					while True:
						try:
							enc_data = self.ssl_out_buff.read()
							if enc_data == b'':
								break
							await self.transport.out_queue.put(enc_data)
						except Exception as e:
							raise e
				except Exception as e:
							raise e

		except asyncio.CancelledError:
			return
		except Exception as e:
			logging.debug('__encrypt_ssl_rec %s' % e)
	
	async def __decrypt_ssl_rec(self):
		"""
		Decrypting the encrypted SSL records from the internal queue and putting them in the queue for the uppr layers to use
		"""
		try:
			while not self.stop_evt.is_set():
				ssl_data = await self.__ssl_in_q.get()
				if ssl_data == b'':
					await self.in_queue.put((ssl_data, None))
					#print('Connection terminated')
					return
				self.ssl_in_buff.write(ssl_data)

				data_buff = b''
				while not self.stop_evt.is_set():
					try:
						data_buff += self.ssl_obj.read()
					except ssl.SSLWantReadError:
						break
					except Exception as e:
						raise e
				#print('data_buff %s' % data_buff)
				if data_buff != b'':
					await self.in_queue.put((data_buff, None))

		except asyncio.CancelledError:
			return
		except Exception as e:
			logging.debug('__decrypt_ssl_rec %s' % e)

	async def __read_ssl_record(self):
		try:
			"""
			buffering the incoming SSL (encrypted) recodrs and putting them in an internal queue one record at a time
			"""
			buffer = b''
			length = None
			while not self.stop_evt.is_set():
				if length is None and len(buffer) >= 6:
					length = int.from_bytes(buffer[3:5], byteorder = 'big', signed = False)
				
				if length is not None and len(buffer) >= length + 5:
					#print('LB raw %s' % len(buffer[:length+5]))
					await self.__ssl_in_q.put(buffer[:length+5])
					buffer = buffer[length+5:]
					length = None
					continue
				
				data, err = await self.transport.in_queue.get()
				if err is not None:
					raise err
				if data == b'':
					await self.__ssl_in_q.put(b'')
					return
				buffer+= data
				

		except asyncio.CancelledError:
			return

		except Exception as e:
			logging.debug('__read_ssl_record %s' % e)
			await self.__ssl_in_q.put(b'')

		#finally:
		#	self.stop_evt.set()
	
	async def __do_handshake(self):
		try:
			ctr = 0
			while not self.stop_evt.is_set():
				ctr += 1
				#print('DST Performing handshake!')
				try:
					self.ssl_obj.do_handshake()
				except ssl.SSLWantReadError:
					#print('DST want %s' % ctr)
					while True:
						client_hello = self.ssl_out_buff.read()
						if client_hello != b'':
							#print('DST client_hello %s' % len(client_hello))
							await self.transport.out_queue.put(client_hello)
						else:
							break
					
					#print('DST wating server hello %s' % ctr)
					server_hello, err = await self.transport.in_queue.get()
					if err is not None:
						raise err
					if server_hello == b'':
						raise Exception('Server closed the connection!')
					#print('DST server_hello %s' % len(server_hello))
					self.ssl_in_buff.write(server_hello)

					continue
				except:
					raise
				else:
					#print('DST handshake ok %s' % ctr)
					#server_fin = tls_out_buff.read()
					#print('DST server_fin %s ' %  server_fin)
					#await out_q.put(server_fin)
					break			

			return True, None

		except Exception as e:
			logging.debug('__do_handshake %s' % e)
			return None, e

	@staticmethod
	async def from_transport(transport, ssl_ctx = None, ssl_cert = None, ssl_key = None):
		try:
			tunnel = SSLClientTunnel()
			tunnel.transport = transport
			if ssl_ctx is not None:
				tunnel.ssl_ctx = ssl_ctx
			elif ssl_key is not None:
				raise NotImplementedError()
			else:
				tunnel.ssl_ctx = ssl.create_default_context()
				tunnel.ssl_ctx.check_hostname = False
				tunnel.ssl_ctx.verify_mode = ssl.CERT_NONE
			_, err = await tunnel.setup()
			if err is not None:
				raise err

			return tunnel, None
		except Exception as e:
			return None, e
```

**Comment by hoodmane** [comment]
> got stuck tackling the Rust requirement

Currently for Rust I think we are blocked on https://github.com/rust-lang/rust/issues/93015, see #2081.

**Comment by hoodmane** [comment]
I think the most reasonable approach would be to start with adding openssl as a shared library. If that works, then we could try to add the Python ssl library too, probably unvendored from the main bundle so people don't have to load openssl if they don't need crypto.

**Comment by hoodmane** [comment]
Anyways @skelsec, if you could open a PR with your partial work on openssl that would be helpful.

**Comment by jayasimha-raghavan-unskript** [comment]
@skelsec did you get a chance to open the PR with your partial work to get openssl? @hoodmane what other packages need to be brought in to have the libcrypto support working?

**Comment by hoodmane** [comment]
Based on https://github.com/voltbuilder/openssl-wasm it looks very doable.

**Comment by skelsec** [comment]
I have decided to give `python-mbedtls` a try and it worked with some minor modification. I'm still in progress of asking the author to do some minor fixes. It also has ciphers available (but again some mods still need to be done).  
Couldn't make ssl work in python still, the progress I made with libcrypto led nowhere :(

**Comment by hoodmane** [comment]
> the progress I made with libcrypto led nowhere

What specific problems did you run into?

**Comment by skelsec** [comment]
![eeeehhhhhpyopen](https://user-images.githubusercontent.com/19204702/157534680-8f1f7a2f-5a57-4a22-8808-10f2e50d3c5d.png)
Everything compiled, openssl+python,  Setup.local was modified to enable SSL pointing to the correct directory. It looks like symbols can't be found for some reason.

**Comment by hoodmane** [comment]
You are talking about the CPython stdlib ssl library? Doesn't that run into trouble with missing sockets?

**Comment by skelsec** [comment]
Yes I'm talking about the built-in ssl lib, since if I need to use a 3rd party module I already found one which is working (2 days ago). I haven't seen errors relating to sockets during compilation.

**Comment by hoodmane** [comment]
Can you share what you did with ssl? I have a fair amount of experience now debugging dynamic linking problems. For instance, does the ssl module try to `dlopen` libcrypto? Otherwise I would worry that if you statically link in ssl that it might not know where to find symbols from a dynamic libcrypto library.

**Comment by hoodmane** [comment]
Closed by #2263.
